sdk/core/core/src/az_http_request_builder.c

```diff
@@ -3,48 +3,271 @@
+    }
+
+    if (p_hrb->headers_end == 0) {
+      new_header_start = headers_start;
+    } else {
+      uint16_t nheader = 1;
+      for (size_t i = 0; i < headers_space - separator_sizes; ++i) {
+        if (headers_start[i] == '\0' && headers_start[i + 1] == '\0') {
+          assert(i < headers_space);
+          assert(i + 1 < headers_space);
+          assert(i + 2 < headers_space);
+
+          if (nheader >= p_hrb->headers_end) {
+            new_header_start = headers_start + (i + sizeof((uint8_t)'\0')) + 1;
+            break;
+          }
+
+          ++nheader;
+        }
+      }
+    }
+
+    size_t const available_space = buffer_end - new_header_start;
+
+    size_t const name_and_value_size = name.size + value.size;
+    size_t const required_space = name_and_value_size + separator_sizes;
+
+    if (name_and_value_size < name.size || name_and_value_size < value.size
+        || required_space < name_and_value_size || required_space > available_space) {
+      return AZ_ERROR_BUFFER_OVERFLOW;
+    }
+
+    az_const_span const new_header_span = { .begin = new_header_start, .size = required_space };
+    if (az_const_span_is_overlap(new_header_span, name)
+        || az_const_span_is_overlap(new_header_span, value)) {
+      return AZ_ERROR_ARG;
+    }
+  }
+
+  assert(new_header_start != buffer_end);
+
+  memcpy(new_header_start, name.begin, name.size);
+  new_header_start += name.size;
+
+  *new_header_start = '\0';
+  ++new_header_start;
+
+  memcpy(new_header_start, value.begin, value.size);
+  new_header_start += value.size;
+
+  new_header_start[0] = '\0';
+  new_header_start[1] = '\0';
+
+  ++(p_hrb->headers_end);
+
   return AZ_OK;
 }
 
 az_result az_http_request_builder_mark_retry_headers_start(az_http_request_builder * const p_hrb) {
-  (void)p_hrb;
+  p_hrb->retry_headers_start = p_hrb->headers_end;
   return AZ_OK;
 }
 
 az_result az_http_request_builder_remove_retry_headers(az_http_request_builder * const p_hrb) {
-  (void)p_hrb;
+  p_hrb->headers_end = p_hrb->retry_headers_start;
   return AZ_OK;
 }
+
+az_result az_http_request_builder_get_header(
+    az_http_request_builder * const p_hrb,
+    uint16_t const index,
+    az_pair * const out_result) {
+  AZ_CONTRACT_ARG_NOT_NULL(p_hrb);
+  AZ_CONTRACT_ARG_NOT_NULL(out_result);
+
+  if (index >= p_hrb->headers_end) {
+    return AZ_ERROR_ARG;
+  }
+
+  uint8_t * const buffer_end = p_hrb->buffer.begin + p_hrb->buffer.size;
+  uint8_t * const headers_start = p_hrb->url.begin + p_hrb->max_url_size;
+
+  size_t const headers_space = buffer_end - headers_start;
+
+  if (headers_start >= buffer_end || headers_space < (2 * sizeof((uint8_t)'\0'))) {
+    return AZ_ERROR_ARG;
+  }
+
+  uint16_t nheader = 0;
+  uint8_t * last_header_start = headers_start;
+  uint8_t * last_header_name_end = last_header_start;
+  for (size_t i = 0; i < headers_space - (2 * sizeof((uint8_t)'\0')); ++i) {
+    if (headers_start[i] == '\0')
+      if (headers_start[i + 1] != '\0') {
```

Comment:
I would like to have positive conditions as a guideline...  (like `headers_start[i + 1] == SEPARATOR)  makes reading code simpler IMO,,, but I am open to opinions here...  @sergey-shandar

Reply:
This no longer applies since we store pairs and not zero character separated values.